src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -82247,6 +82235,69 @@ static void F(object o)
             comp.VerifyTypes();
         }
 
+        /// <summary>
+        /// Nullability of variable members is tracked up to a fixed depth.
+        /// </summary>
+        [Fact]
+        [WorkItem(31395, "https://github.com/dotnet/roslyn/issues/31395")]
+        public void InheritNullabilityMaxDepth_01()
+        {
+            var source =
+@"#pragma warning disable 8618
+class A
+{
+    internal B? B;
+}
+class B
+{
+    internal A? A;
+}
+class Program
+{
+    static void F()
+    {
+        var a1 = new A() { B = new B() };
+        var a2 = new A() { B = new B() { A = a1 } };
+        var a3 = new A() { B = new B() { A = a2 } };
+        a1.B.ToString();
+        a2.B.A.B.ToString();
+        a3.B.A.B.A.B.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (19,9): warning CS8602: Dereference of a possibly null reference.
+                //         a3.B.A.B.A.B.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "a3.B.A.B.A.B").WithLocation(19, 9));
+        }
+
+        /// <summary>
+        /// Nullability of variable members is tracked up to a fixed depth.
+        /// </summary>
+        [Fact]
+        [WorkItem(31395, "https://github.com/dotnet/roslyn/issues/31395")]
+        public void InheritNullabilityMaxDepth_02()
+        {
+            var source =
+@"class Program
+{
+    static void F(string x, object y)
+    {
+        (((((string? x5, object? y5) x4, string? y4) x3, object? y3) x2, string? y2) x1, object? y1) t = (((((x, y), x), y), x), y);
+        t.y1.ToString();
+        t.x1.y2.ToString();
+        t.x1.x2.y3.ToString();
+        t.x1.x2.x3.y4.ToString();
+        t.x1.x2.x3.x4.y5.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         t.x1.x2.x3.x4.y5.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t.x1.x2.x3.x4.y5").WithLocation(10, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider adding tests we discussed if not already covered: `while (b) { x = x.F; }` and `while(b) { x.F = x; }`?

Reply:
We have a number of similar cycle tests: see `Members_FieldCycle_*`. I believe those cases are covered.

---
In reply to: [284858771](https://github.com/dotnet/roslyn/pull/35720#discussion_r284858771) [](ancestors = 284858771)